rust/connlib/clients/shared/src/eventloop.rs

```diff
@@ -0,0 +1,570 @@
+use crate::{
+    messages::{
+        BroadcastGatewayIceCandidates, Connect, ConnectionDetails, EgressMessages,
+        GatewayIceCandidates, IngressMessages, InitClient, RemoveResource, ReplyMessages,
+    },
+    PHOENIX_TOPIC,
+};
+use anyhow::Result;
+use async_compression::tokio::bufread::GzipEncoder;
+use bimap::BiMap;
+use connlib_shared::{
+    messages::{
+        ConnectionAccepted, DnsServer, GatewayId, GatewayResponse, IpDnsServer, ResourceAccepted,
+        ResourceId,
+    },
+    Callbacks, IpProvider,
+};
+use firezone_tunnel::ClientTunnel;
+use ip_network::IpNetwork;
+use phoenix_channel::{ErrorReply, OutboundRequestId, PhoenixChannel};
+use reqwest::header::{CONTENT_ENCODING, CONTENT_TYPE};
+use std::{
+    collections::HashMap,
+    convert::Infallible,
+    io,
+    net::IpAddr,
+    path::PathBuf,
+    str::FromStr,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::{
+    io::BufReader,
+    time::{Instant, Interval, MissedTickBehavior},
+};
+use tokio_util::codec::{BytesCodec, FramedRead};
+use url::Url;
+
+const DNS_PORT: u16 = 53;
+const DNS_SENTINELS_V4: &str = "100.100.111.0/24";
+const DNS_SENTINELS_V6: &str = "fd00:2021:1111:8000:100:100:111:0/120";
+
+pub struct Eventloop<C: Callbacks> {
+    tunnel: ClientTunnel<C>,
+    tunnel_init: bool,
+
+    portal: PhoenixChannel<(), IngressMessages, ReplyMessages>,
+    connection_intents: SentConnectionIntents,
+    log_upload_interval: tokio::time::Interval,
+}
+
+impl<C: Callbacks> Eventloop<C> {
+    pub(crate) fn new(
+        tunnel: ClientTunnel<C>,
+        portal: PhoenixChannel<(), IngressMessages, ReplyMessages>,
+    ) -> Self {
+        Self {
+            tunnel,
+            portal,
+            tunnel_init: false,
+            connection_intents: SentConnectionIntents::default(),
+            log_upload_interval: upload_interval(),
+        }
+    }
+}
+
+impl<C> Eventloop<C>
+where
+    C: Callbacks + 'static,
+{
+    #[tracing::instrument(name = "Eventloop::poll", skip_all, level = "debug")]
+    pub fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Infallible, phoenix_channel::Error>> {
+        loop {
+            match self.tunnel.poll_next_event(cx) {
+                Poll::Ready(Ok(event)) => {
+                    self.handle_tunnel_event(event);
+                    continue;
+                }
+                Poll::Ready(Err(e)) => {
+                    tracing::error!("Tunnel failed: {e}");
+                    continue;
+                }
+                Poll::Pending => {}
+            }
```

Comment:
Tunnel errors **do not** end up calling `onDisconnect`. They are just logged here. This is also what we currently do on `main`. In the future, I'd like to make a type-safe error API where only very specific errors can be returned from `poll_next_event` and those will then always call `onDisconnect` to actually shut things down.